### PR TITLE
Enforce semver during build

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Enhancement / new feature

### Description

With this change the build will fail if we make changes that do not conform to the semantic versioning rules.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by SamBarker** [comment]
We should also update the release process to update the version used for providing the API baseline. I think however that is a separate concern to this PR. Especially as its not, to me at least, clear when we should change the dependency. 

Should a release of 0.5.1 be used as the API baseline or should we stick to 0.5.0 until 0.6.0 is released? 

Tempted to say doing the simple thing is good enough that we bump it on every release but would like to hear others thoughts.

File: kroxylicious-api/pom.xml [hunk 1]
```diff
@@ -179,6 +181,40 @@
...
+                        <!-- specify the old version directly so its deterministic -->
+                        <dependency>
+                            <groupId>${project.groupId}</groupId>
+                            <artifactId>${project.artifactId}</artifactId>
+                            <version>${ApiCompatability.ReferenceVersion}</version>
+                            <type>jar</type>
+                        </dependency>
+                    </oldVersion>
+                    <parameter>
+                        <accessModifier>public</accessModifier>
```
Comment (robobario) [inline kroxylicious-api/pom.xml:199]: What's this one do? "Sets the access modifier level (public, protected, package_protected, private)." I guess it's this from the introduction: `Per default private and package protected classes and class members are not compared. If necessary, the access modifier of the classes and class members to be compared can be set to public, protected, package or private.`. What is the default? Should we check protected members too?

Comment (SamBarker) [reply kroxylicious-api/pom.xml:199]: IIRC I copied it from the Flink config and didn't think too hard about it. My theory being we only care about the public API for semver. However you make an interesting point about protected. I wonder if we actually use protected in the API.

IIRC the default is `public` but I would need to confirm.

**Comment by robobario** [comment]
As an experiment I tried deleting a public constructor. 

```
+++ b/kroxylicious-api/src/main/java/io/kroxylicious/proxy/plugin/UnknownPluginInstanceException.java
@@ -14,8 +14,4 @@ public class UnknownPluginInstanceException extends RuntimeException {
     public UnknownPluginInstanceException(String message) {
         super(message);
     }
-
-    public UnknownPluginInstanceException(String message, Throwable cause) {
-        super(message, cause);
-    }
 }
```

Which causes the build to fail for semver compliance, fair enough. But then I bump the project version to 1.0.0-SNAPSHOT and it still fails with:

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.18.0:cmp (default) on project kroxylicious-api: There is at least one incompatibility: io.kroxylicious.proxy.plugin.UnknownPluginInstanceException.UnknownPluginInstanceException(java.lang.String,java.lang.Throwable):CONSTRUCTOR_REMOVED -> [Help 1]
```

I get the feeling that:
```breakBuildOnBinaryIncompatibleModifications	true	false	If set to true, the build breaks in case a binary incompatible modification has been detected.
```
isn't concerned with the version at all and will fail on any binary incompatibility.

Maybe we want it set up with `breakBuildBasedOnSemanticVersioning: true` and `breakBuildBasedOnSemanticVersioningForMajorVersionZero: true`, `breakBuildOnBinaryIncompatibleModifications: false`. I tested with that and when I remove the constructor in 0.6.0-SNAPSHOT I get:

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.18.0:cmp (default) on project kroxylicious-api: Versions of archives indicate a minor change but binary incompatible changes found. -> [Help 1]
```

Then I set the versions to 1.0.0-SNAPSHOT and the build passes.

**Comment by robobario** [comment]
Thanks Sam, I'm keen to try being harsher with kroxylicious-api changes since it's for external consumption. 

One thing I'm thinking of is we have 9 Deprecations in kroxylicious-api currently. Do we want to carry them until Kroxylicious 1.0.0? We could also consider going back to a separate version/cadence for kroxylicious-api but I think we backed off that because it added complexity to the release.

**Comment by SamBarker** [comment]
>One thing I'm thinking of is we have 9 Deprecations in kroxylicious-api currently. Do we want to carry them until Kroxylicious 1.0.0? We could also consider going back to a separate version/cadence for kroxylicious-api but I think we backed off that because it added complexity to the release.

I think we give them a minor or maybe 2 minor releases until we hit 1.0. 

Yeah we moved away because the release complexity wasn't worth the benefits of the separate cadence. Especially as we didn't have a separate cadence. I think its solveable but I'm un-convinced its worth the effort.

**Review by robobario — APPROVED** [review]
LGTM, lets try it out

**Comment by tombentley** [comment]
While we're in `0.x.y` I think:

* it's OK if `0.x₂,y` is not binary or source compatible with `0.x₁.y` (where `x₁ < x₂`). I.e. it's OK for us to break compatibility between minor versions.
* it's not OK if `0.x.y₂` is not compatible `0.x.y₁` for any `y₁ < y₂`. I.e. it's not OK for us to break compatibility between patch versions.

If that's what this is implementing then let's merge this PR and see how we get on with it in practice.

**Comment by SamBarker** [comment]
Its slightly stricter check. Its enforcing 1.X.Y+ semver rules (with an escape hatch if we decide to make a breaking change). If I add a random boolean to `FilterContext.eventLoop` i.e. `FilterContext.eventLoop(boolean ignored)` 
it fails with 

```
[INFO] --- japicmp-maven-plugin:0.18.0:cmp (default) @ kroxylicious-api ---
[INFO] Written file '/Users/sbarker/src/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.diff'.
[INFO] Written file '/Users/sbarker/src/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.xml'.
[INFO] Written file '/Users/sbarker/src/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.html'.
Apr 15, 2024 4:33:06 PM japicmp.output.incompatible.IncompatibleErrorOutput warn
WARNING: Incompatibility detected: Requires semantic version level MAJOR: JApiMethod [oldMethod=io.kroxylicious.proxy.filter.FilterFactoryContext.eventLoop(), newMethod=n.a., returnType=JApiReturnType [oldReturnTypeOptional=java.util.concurrent.ScheduledExecutorService, newReturnTypeOptional=value absent, changeStatus=REMOVED], getCompatibilityChanges()=[METHOD_REMOVED]]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Parent POM 0.6.0-SNAPSHOT:
[INFO] 
[INFO] Parent POM ......................................... SUCCESS [  4.686 s]
[INFO] kroxylicious-annotations ........................... SUCCESS [  5.384 s]
[INFO] kroxylicious-kms ................................... SUCCESS [  4.718 s]
[INFO] kroxylicious-kms-test-support ...................... SUCCESS [  3.693 s]
[INFO] Code generation plugin ............................. SUCCESS [ 10.656 s]
[INFO] Kroxylicious API ................................... FAILURE [ 16.678 s]
[INFO] In-memory KMS ...................................... SKIPPED
[INFO] kroxylicious-kms-provider-kroxylicious-inmemory-test-support SKIPPED
[INFO] Hashicorp Vault KMS ................................ SKIPPED
[INFO] kroxylicious-kms-provider-hashicorp-vault-test-support SKIPPED
[INFO] Bill of Materials .................................. SKIPPED
[INFO] Filter unit test support ........................... SKIPPED
[INFO] Proxy runtime ...................................... SKIPPED
[INFO] Integration test support ........................... SKIPPED
[INFO] Filter implementations ............................. SKIPPED
[INFO] Multitenancy filter ................................ SKIPPED
[INFO] Record validation filter ........................... SKIPPED
[INFO] Simple filters ..................................... SKIPPED
[INFO] Kroxylicious Encryption ............................ SKIPPED
[INFO] Proxy application .................................. SKIPPED
[INFO] Sample filters ..................................... SKIPPED
[INFO] Integration tests .................................. SKIPPED
[INFO] kroxylicious-systemtests ........................... SKIPPED
[INFO] Micro benchmarks ................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  46.216 s
[INFO] Finished at: 2024-04-15T16:33:06+12:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.18.0:cmp (default) on project kroxylicious-api: Versions of archives indicate a minor change but binary incompatible changes found. -> [Help 1]
```

If we intentionally break compatibility we need to set `ApiCompatability.EnforceForMajorVersionZero` to `false`.

**Pull request merged**